gnomad_constraint/utils/constraint.py

```diff
@@ -196,11 +199,240 @@ def create_constraint_training_dataset(
+        Default is None.
+    :param cov_cutoff: Median coverage cutoff. Sites with coverage above this cutoff
+        are considered well covered and will be used to build plateau models. Sites
+        below this cutoff have low coverage and will be used to build coverage models.
+        Default is `COVERAGE_CUTOFF`.
+    :return: Table with `expected_variants` (expected variant counts) and `obs_exp`
+        (observed:expected ratio) annotations.
+    """
+    # Add necessary constraint annotations for grouping
+    if custom_model == "worst_csq":
+        (
+            context_ht,
+            _,
+        ) = annotate_constraint_groupings(context_ht, "worst_csq_by_gene")
+        exome_ht, grouping = annotate_constraint_groupings(
+            exome_ht, "worst_csq_by_gene"
+        )
+
+    else:
+        context_ht, _ = annotate_constraint_groupings(
+            context_ht, "transcript_consequences"
+        )
+        exome_ht, grouping = annotate_constraint_groupings(
+            exome_ht, "transcript_consequences"
+        )
+
+    # Allele frequency information for high-quality genotypes (GQ >= 20; DP >= 10; and
+    # AB >= 0.2 for heterozygous calls) in all release samples in gnomAD.
+    freq_expr = exome_ht.freq[0]
+
+    # Set up the criteria to exclude variants not observed in the dataset, low-quality
+    # variants, and variants with allele frequency above the `max_af` cutoff.
+    keep_criteria = (
+        (freq_expr.AC > 0)
+        & exome_ht.pass_filters
+        & (freq_expr.AF <= max_af)
+        & (exome_ht.coverage > 0)
+    )
+    keep_annotations += grouping
+
+    # Keep variants that satisfy the criteria above.
+    # Count the observed variants in the entire Table and in each downsampling grouped
+    # by `keep_annotations`.
+    observed_ht = count_variants_by_group(
+        exome_ht.filter(keep_criteria).select(*list(keep_annotations) + ["freq"]),
+        additional_grouping=grouping,
+        partition_hint=partition_hint,
+        count_downsamplings=pops,
+        use_table_group_by=True,
+    )
+    observed_ht = observed_ht.transmute(observed_variants=observed_ht.variant_count)
+
+    # Filter the `exome_ht` to rows that don’t match the criteria above.
+    # Anti join the `context_ht` with filtered `exome_ht`, so that `context_ht` only
+    # has rows that match the criteria above in the `exome_ht` or are never in
+    # the `exome_ht`.
+    context_ht = context_ht.select(*keep_annotations).anti_join(
+        exome_ht.filter(keep_criteria, keep=False)
+    )
+
+    # Count the possible variants in the context Table grouped by `keep_annotations`.
+    possible_ht = count_variants_by_group(
+        context_ht,
+        additional_grouping=grouping,
+        partition_hint=partition_hint,
+        use_table_group_by=True,
+    )
+    possible_ht = annotate_with_mu(possible_ht, mutation_ht)
+    possible_ht = possible_ht.transmute(possible_variants=possible_ht.variant_count)
```

Comment:
Change to use the suggested function above.
```suggestion
    ht = create_observed_and_possible_ht(
        exome_ht,
        context_ht,
        mutation_ht,
        max_af,
        keep_annotations,
        pops,
        grouping,
        partition_hint,
        filter_coverage_over_0=True,
        filter_to_canonical_synonymous=False,
    )
```